platform/osgi/platform-osgi-configadmin/src/main/java/org/codice/felix/cm/file/DelegatingPersistenceManager.java

```diff
@@ -0,0 +1,112 @@
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.felix.cm.file;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Dictionary;
+import java.util.List;
+import org.apache.felix.cm.PersistenceManager;
+import org.codice.felix.cm.internal.ConfigurationContextFactory;
+import org.codice.felix.cm.internal.ConfigurationPersistencePlugin;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.util.tracker.ServiceTracker;
+
+/**
+ * The {@link DelegatingPersistenceManager} is responsible for iterating over relevant configuration
+ * plugin points prior to moving the configs to the inner {@link PersistenceManager}.
+ */
+public class DelegatingPersistenceManager extends WrappedPersistenceManager {
+
+  private final ServiceTracker configPersistenceTracker;
+
+  private final ServiceRegistration<ConfigurationContextFactory> contextFactoryRegistration;
+
+  public DelegatingPersistenceManager(PersistenceManager persistenceManager) {
+    this(
+        persistenceManager,
+        new ServiceTracker(getBundleContext(), ConfigurationPersistencePlugin.class, null),
+        getBundleContext()
+            .registerService(
+                ConfigurationContextFactory.class, new ConfigurationContextFactoryImpl(), null));
+  }
+
+  DelegatingPersistenceManager(
+      PersistenceManager persistenceManager,
+      ServiceTracker tracker,
+      ServiceRegistration<ConfigurationContextFactory> registration) {
+    super(persistenceManager);
+    tracker.open();
+
+    this.configPersistenceTracker = tracker;
+    this.contextFactoryRegistration = registration;
+  }
+
+  @Override
+  public void close() throws Exception {
+    configPersistenceTracker.close();
+    if (contextFactoryRegistration != null) {
+      contextFactoryRegistration.unregister();
+    }
+    super.close();
+  }
+
+  @Override
+  public void store(String pid, Dictionary properties) throws IOException {
+    ConfigurationContextImpl context = createContext(pid, properties);
+    if (context.shouldBeVisibleForProcessing()) {
+      List<ConfigurationPersistencePlugin> plugins =
+          asPlugins(configPersistenceTracker.getServices());
+      for (ConfigurationPersistencePlugin plugin : plugins) {
+        plugin.handleStore(context);
+      }
+    }
+    super.store(pid, properties);
+  }
+
+  @Override
+  public void delete(String pid) throws IOException {
+    List<ConfigurationPersistencePlugin> plugins =
+        asPlugins(configPersistenceTracker.getServices());
+    for (ConfigurationPersistencePlugin plugin : plugins) {
+      plugin.handleDelete(pid);
```

Comment:
✏️ With an update to the `ConfigurationPersistencePlugin` interface such that it throws `UncheckedIOException` instead of the checked version, these methods could then be updated to simply call `forEach` on a returned stream from an updated `asPlugins` method. It's almost a stylistic suggestion more than anything else, as there aren't going to be so many plugins that constructing the List from the array will be an expensive operation.

Reply:
@coyotesqrl Aren't you then forced to unwrap this UncheckedIOException with a try/catch around the foreach? You don't want that to bubble out.
I am a fan of streams and forEach() but not in these circumstances.

Reply:
If the desire is to avoid the runtime exception, then yes. We'd need to wrap the call on the stream in a try/catch. It's not worth the effort just to avoid creation of the temporary collection.

Reply:
On the other end ... I hate creating temporary collections :-)

Reply:
Note also this is the config admin bundle and we cannot use Java 8 constructs.

Reply:
comment to that regard in the file